group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Utils.java

```diff
@@ -235,4 +238,20 @@ public static ApiMessage messageOrNull(ApiMessageAndVersion apiMessageAndVersion
             return apiMessageAndVersion.message();
         }
     }
+
+    public static long hashSubscriptionMetadata(Map<String, TopicMetadata> subscriptionMetadata) {
```

Comment:
I would like to add some implementation detail about the "cache" in GroupMetadataManager, so we don't waste resource to store unsubscribed topics or recalculate a topic hash with same value.

1. Only compute topic hash which is subscribed by a group. That means we will compute topic hash in `GroupMetadataManager#subscribeGroupToTopic` when it's first time subscribed by one of groups. Also, the hash value will be removed in `GroupMetadataManager#unsubscribeGroupFromTopic` if there is no group subscribes it.
2. In `GroupMetadataManager#onNewMetadataImage`, we only recalculate topic hash which is in `topicsDelta.changedTopics()`.

Reply:
Why not using a single hash? The issue is that we don't always know which topics have changed (e.g. after a coordinator failover) so we have to recompute the cache anyway.

> The coordinator will maintain an in-memory cache Map<UUID, Integer> where the key is the topic UUID and the value is the hash based on 'name + partitions + racks'.

If we store a hash per topic, it should be stored by topic name because member subscribe by topic name and we must also catch the topic recreation.

> The cache will be updated whenever a group needs to calculate the hash for an uncached topic or when new metadata brings a topic change.

Could you elaborate on how this will work?

Reply:
@FrankYang0529 
Is he saying that whenever [`consumerGroupHeartbeat`](https://github.com/frankyang0529/kafka/blob/2ff4c89b8e88ec1170ab698cb016e33564724909/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L1796-L1813) is called, it’s necessary to recompute the hash for the incoming `<groupName, topicMetadata>` map? This would mean recalculating the hash for either each group name’s individual metadata or for the entire map, which is indeed quite time-consuming.

Reply:
> Why not using a single hash?

If just using a single hash for subscribed topics in a group, that means each group needs to recalculate the hash value. If using a hash for a topic, the coordinator can calculate a new hash when there is topic update. each group can get the latest hash and compare with current hash to know whether there is any update.

![image](https://github.com/user-attachments/assets/4bf67cd0-20d6-4a50-90b3-57f7e24025b8)

> The cache will be updated whenever a group needs to calculate the hash for an uncached topic or when new metadata brings a topic change.

Like I mentioned in previous comment (https://github.com/apache/kafka/pull/17444#discussion_r1820720961). When a coordinator starts, it replays all records. At that time, the coordinator calculate initial hash in `GroupMetadataManager#subscribeGroupToTopic`. When there is new metadata image, we calculate a new hash for each changed topic and request metadata refresh for related groups. In next group heartbeat, the coordinator constructs a new hash map from the cache and compares with current hash map in the group.

Reply:
> Why not using a single hash? The issue is that we don't always know which topics have changed (e.g. after a coordinator failover) so we have to recompute the cache anyway.

As @FrankYang0529 explained, we calculate the hash of topics as needed. For example:
1. We can rebuild the cache for all subscribed topics during a metadata update.
2. Alternatively, the calculation can occur when we need to call group.computeSubscriptionMetadata, and we remove the topic hash from the cache if the topics have changed during a metadata update.

By the way, here are the approaches we've discussed:
1. TopicMetadata: Not optimal for space.
2. Single Hash: Not efficient for computing, as a change in one topic could trigger all related groups to recalculate the hash.
3. Hash per Topic: Balances space and computation by using a cache.

Reply:
Thanks. Let me think about it for a while. I do have one first comment though. When we replay records, we should only replay the persisted state. We should not compute hashes there based on the metadata image there.

Reply:
> When we replay records, we should only replay the persisted state. We should not compute hashes there based on the metadata image there.

There are two parts of topic hash map. One is in coordinator (GroupMetadataManager), and another is in group (ModernGroup).

For the coordinator, it's in memory cache. When a coordinator starts, we will recalculate it from metadata image in `GroupMetadataManager#subscribeGroupToTopic`. When there is a new metadata image, we recalculate changed topic hash.

For the group, the hash map is updated when we replay `ConsumerGroupPartitionMetadataValue`. We will add a new field `hash` to `TopicMetadata`. Although it still calls `TopicMetadata`, we only cares hash value, so in the code, we only need to maintain `Map<String, Long> subscribedTopicHash`.

Reply:
> For the coordinator, it's in memory cache. When a coordinator starts, we will recalculate it from metadata image in GroupMetadataManager#subscribeGroupToTopic

This might be too detailed, but perhaps we don't need to build the hash cache at this stage. Instead, we could build the cache in `GroupMetadataManager#consumerGroupHeartbeat` when it's actually needed. For example:

```java
    public Map<String, Integer> computeSubscriptionMetadata(
        Map<String, Integer> subscribedTopicNames,
        TopicsImage topicsImage,
        Map<String, Integer> cache
    ) {
        // Create the topic metadata for each subscribed topic.
        Map<String, Integer> newSubscriptionMetadata = new HashMap<>(subscribedTopicNames.size());

        subscribedTopicNames.forEach((topicName, count) -> {
            TopicImage topicImage = topicsImage.getTopic(topicName);
            if (topicImage != null) {
                var topicMetadata = new TopicMetadata(
                        topicImage.id(),
                        topicImage.name(),
                        topicImage.partitions().size()
                );
                newSubscriptionMetadata.put(topicName, cache.computeIfAbsent(topicName, __ -> topicMetadata.hashCode()));
            }
        });
        return Collections.unmodifiableMap(newSubscriptionMetadata);
    }
```

Reply:
Hi @chia7712 / @dajac, I updated [KIP-1101](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1101%3A+Trigger+rebalance+on+rack+topology+changes) to include the method we mention in this thread. Please take a look when you have time. Thank you.

Reply:
@FrankYang0529 Please take a look at the above comment. We don’t need to update the topic hash in onNewMetadataImage since lazy evaluation is more efficient. Otherwise, we may waste computation if there are multiple metadata events without any heartbeat events.

Additionally, we should evict the topic from the cache if the topic is either changed or deleted. This ensures that the subsequent computeSubscriptionTopicHash generates an up-to-date hash.

Reply:
Hi @chia7712, thanks for the great suggestion. Updated KIP.